File: mobile/android/app/src/main/java/com/termirust/mobile/controller/TerminalAcceptance.kt
package com.termirust.mobile.controller

import kotlin.math.floor

data class TerminalAcceptanceLayout(
    val fontSize: Double,
    val displayedFontSize: Double,
    val columns: Int,
    val rows: Int,
    val compactControls: Boolean,
)

data class TerminalBackgroundDecision(
    val releaseWriter: Boolean,
    val coverPrivacy: Boolean,
    val clearPendingInput: Boolean,
    val clearPendingResize: Boolean,
)

object TerminalAcceptance {
    const val MINIMUM_FONT_SIZE = 10.0
    const val MAXIMUM_FONT_SIZE = 32.0
    const val MAXIMUM_ACCESSIBILITY_CHARACTERS = 4_096
    const val MINIMUM_TOUCH_TARGET = 48.0

    fun layout(
        width: Double,
        height: Double,
        requestedFontSize: Double,
        textScale: Double,
    ): TerminalAcceptanceLayout {
        val fontSize = requestedFontSize.coerceIn(MINIMUM_FONT_SIZE, MAXIMUM_FONT_SIZE)
        val scale = textScale.coerceAtLeast(1.0)
        val displayedFontSize = fontSize * scale
        val usableWidth = (width - 24.0).coerceAtLeast(1.0)
        val usableHeight = (height - 24.0).coerceAtLeast(1.0)
        return TerminalAcceptanceLayout(
            fontSize = fontSize,
            displayedFontSize = displayedFontSize,
            columns = floor(usableWidth / (displayedFontSize * 0.62)).toInt().coerceIn(20, 400),
            rows = floor(usableHeight / (displayedFontSize * 1.35)).toInt().coerceIn(5, 200),
            compactControls = width < 600.0 || scale >= 1.6,
        )
    }

    fun accessibleOutput(
        lines: List<String>,
        maximumCharacters: Int = MAXIMUM_ACCESSIBILITY_CHARACTERS,
    ): String {
        if (maximumCharacters <= 0) return ""
        val output = lines.joinToString("\n").trim()
        return if (output.length <= maximumCharacters) output else output.takeLast(maximumCharacters)
    }

    fun backgroundDecision(writerHeld: Boolean) = TerminalBackgroundDecision(
        releaseWriter = writerHeld,
        coverPrivacy = true,
        clearPendingInput = true,
        clearPendingResize = true,
    )
}

File: apps/ios/TermiRustMobile/Terminal/TerminalAcceptance.swift
import Foundation

struct TerminalAcceptanceLayout: Equatable, Sendable {
    let fontSize: Double
    let displayedFontSize: Double
    let columns: Int
    let rows: Int
    let compactControls: Bool
}

struct TerminalBackgroundDecision: Equatable, Sendable {
    let releaseWriter: Bool
    let coverPrivacy: Bool
    let clearPendingInput: Bool
    let clearPendingResize: Bool
}

enum TerminalAcceptance {
    static let minimumFontSize = 10.0
    static let maximumFontSize = 32.0
    static let maximumAccessibilityCharacters = 4_096
    static let minimumTouchTarget = 44.0

    static func layout(
        width: Double,
        height: Double,
        requestedFontSize: Double,
        textScale: Double
    ) -> TerminalAcceptanceLayout {
        let fontSize = requestedFontSize.clamped(to: minimumFontSize...maximumFontSize)
        let scale = max(textScale, 1)
        let displayedFontSize = fontSize * scale
        let usableWidth = max(width - 24, 1)
        let usableHeight = max(height - 24, 1)
        let columns = Int((usableWidth / (displayedFontSize * 0.62)).rounded(.down))
            .clamped(to: 20...400)
        let rows = Int((usableHeight / (displayedFontSize * 1.35)).rounded(.down))
            .clamped(to: 5...200)
        return TerminalAcceptanceLayout(
            fontSize: fontSize,
            displayedFontSize: displayedFontSize,
            columns: columns,
            rows: rows,
            compactControls: width < 600 || scale >= 1.6
        )
    }

    static func accessibleOutput(
        lines: [String],
        maximumCharacters: Int = maximumAccessibilityCharacters
    ) -> String {
        guard maximumCharacters > 0 else { return "" }
        let output = lines.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        guard output.count > maximumCharacters else { return output }
        return String(output.suffix(maximumCharacters))
    }

    static func backgroundDecision(writerHeld: Bool) -> TerminalBackgroundDecision {
        TerminalBackgroundDecision(
            releaseWriter: writerHeld,
            coverPrivacy: true,
            clearPendingInput: true,
            clearPendingResize: true
        )
    }
}

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
